modules/audio/src/main/java/com/gluonhq/attach/audio/impl/AndroidAudioService.java

```diff
@@ -0,0 +1,236 @@
+ * (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL GLUON BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.gluonhq.attach.audio.impl;
+
+import com.gluonhq.attach.audio.Audio;
+import com.gluonhq.attach.audio.AudioService;
+import com.gluonhq.attach.storage.StorageService;
+import com.gluonhq.attach.util.Services;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.util.Optional;
+
+/**
+ * @author Almas Baimagambetov (almaslvl@gmail.com)
+ */
+public class AndroidAudioService implements AudioService {
+
+    static {
+        System.loadLibrary("audio");
+    }
+
+    private File privateStorage;
+
+    @Override
+    public Optional<Audio> loadSound(URL url) {
+        String fullName = prepareFile(url, "sounds/");
+
+        if (fullName.isEmpty())
+            return Optional.empty();
+
+        int audioId = loadSoundImpl(fullName);
+
+        if (audioId == -1)
+            return Optional.empty();
+
+        return Optional.of(new AndroidAudio(audioId));
+    }
+
+    @Override
+    public Optional<Audio> loadMusic(URL url) {
+        String fullName = prepareFile(url, "music/");
+
+        if (fullName.isEmpty())
+            return Optional.empty();
+
+        int audioId = loadMusicImpl(fullName);
+
+        if (audioId == -1)
+            return Optional.empty();
+
+        return Optional.of(new AndroidAudio(audioId));
+    }
+
+    /**
+     * Copy file (if it doesn't exist) to private storage.
+     *
+     * @param url where the file is
+     * @param subDirName subdirectory name where the file will go
+     * @return file name in private storage if file was copied or if it already exists at destination,
+     * or empty String if there are any errors during copying
+     */
+    private String prepareFile(URL url, String subDirName) {
```

Comment:
Add some checks on `url` and `subdirName`, both can't be null, can they?

Reply:
`url` can only be null if user passed it. I'll add a check but to public API, not internal. `subDirName` only used internally, so never null.